include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -175,100 +174,75 @@ make_iter_mode(const _Iterator& __it) -> decltype(iter_mode<outMode>()(__it))
-#else
-        sycl::program __program(__exec.queue().get_context());
-
-        __program.build_with_kernel_type<_DerivedKernelName>();
-        return __program.get_kernel<_DerivedKernelName>();
-#endif
-    }
-};
-
-template <typename... _Name>
-class __parallel_for_kernel : public __kernel_name_base<__parallel_for_kernel<_Name...>>
-{
-};
 template <typename... _Name>
-class __parallel_reduce_kernel : public __kernel_name_base<__parallel_reduce_kernel<_Name...>>
-{
-};
-template <typename... _Name>
-class __parallel_scan_local_kernel : public __kernel_name_base<__parallel_scan_local_kernel<_Name...>>
-{
-};
+class __parallel_reduce_kernel;
+
 template <typename... _Name>
-class __parallel_scan_global_kernel : public __kernel_name_base<__parallel_scan_global_kernel<_Name...>>
-{
-};
+class __parallel_scan_local_kernel;
+
 template <typename... _Name>
-class __parallel_scan_propagate_kernel : public __kernel_name_base<__parallel_scan_propagate_kernel<_Name...>>
-{
-};
+class __parallel_scan_group_kernel;
+
 template <typename... _Name>
-class __parallel_find_or_kernel : public __kernel_name_base<__parallel_find_or_kernel<_Name...>>
-{
-};
+class __parallel_scan_propagate_kernel;
+
 template <typename... _Name>
-class __parallel_merge_kernel : public __kernel_name_base<__parallel_merge_kernel<_Name...>>
-{
-};
+class __parallel_find_or_kernel;
+
 template <typename... _Name>
-class __parallel_sort_leaf_kernel : public __kernel_name_base<__parallel_sort_leaf_kernel<_Name...>>
-{
-};
+class __parallel_sort_leaf_kernel;
+
 template <typename... _Name>
-class __parallel_sort_global_kernel : public __kernel_name_base<__parallel_sort_global_kernel<_Name...>>
-{
-};
+class __parallel_sort_global_kernel;
+
 template <typename... _Name>
-class __parallel_sort_copy_back_kernel : public __kernel_name_base<__parallel_sort_copy_back_kernel<_Name...>>
-{
-};
+class __parallel_sort_copy_back_kernel;
 
 //------------------------------------------------------------------------
 // parallel_for - async pattern
 //------------------------------------------------------------------------
+template <typename _KernelName>
+struct __parallel_for_submitter;
+
+template <typename... _Name>
+struct __parallel_for_submitter<__internal::__optional_kernel_name<_Name...>>
+{
+    template <typename _ExecutionPolicy, typename _Fp, typename _Index, typename... _Ranges>
+    __future<void>
+    operator()(_ExecutionPolicy&& __exec, _Fp __brick, _Index __count, _Ranges&&... __rngs) const
+    {
+        assert(oneapi::dpl::__ranges::__get_first_range_size(__rngs...) > 0);
+        _PRINT_INFO_IN_DEBUG_MODE(__exec);
+        auto __event = __exec.queue().submit([&__rngs..., &__brick, __count](sycl::handler& __cgh) {
+            //get an access to data under SYCL buffer:
+            oneapi::dpl::__ranges::__require_access(__cgh, __rngs...);
+
+            __cgh.parallel_for<_Name...>(sycl::range</*dim=*/1>(__count), [=](sycl::item</*dim=*/1> __item_id) {
```

Comment:
Why do we use the "such approach" like `__cgh.parallel_for<_Name...>` for not all "patterns", just for parallel_for, sort, reduce, merge? Or other words, we do we have different "name usage model"? in some case a user m.b. sure that his kernel name is passed in the other cases (reduce, scan..) we compose name... (partially,, I understand that we need calc wg size.... but in any case it looks inconsistency form code perspective and form user' usage model perspective).

Reply:
Yes, I agree with inconsistency but we have to do that because we don't want to rely on built-in where we don't have mandatory kernel name, and on the other hand we don't have another reliable way to get work group size other than with kernel name and built-in